go-update/src/main/java/org/reactome/release/goupdate/GoTermsUpdater.java

```diff
@@ -0,0 +1,545 @@
+			}
+			else if (termStarted)
+			{
+				currentGOID = GoLineProcessor.processLine(line, currentGOID, goTermsFromFile);
+			}
+		}
+		
+		// Now process all the goTerms.
+		for (String goID : goTermsFromFile.keySet())
+		{
+			GoTermInstanceModifier goTermModifier;
+			GONamespace currentCategory = (GONamespace) goTermsFromFile.get(goID).get(GoUpdateConstants.NAMESPACE);
+			//String currentDefinition = (String) goTerms.get(goID).get(GoUpdateConstants.DEF);
+			// Now we need to process the Term that was just finished.
+			List<GKInstance> goInstances = allGoInstances.get(goID);
+			// First let's make sure the GO Term is not obsolete.
+			if ( !goTermsFromFile.get(goID).containsKey(GoUpdateConstants.IS_OBSOLETE) && !goTermsFromFile.get(goID).containsKey(GoUpdateConstants.PENDING_OBSOLETION))
+			{
+				if (goInstances==null)
+				{
+					// Create a new Instance if there is nothing in the current list of instances.
+					goTermModifier = new GoTermInstanceModifier(this.adaptor, this.instanceEdit);
+					Long dbID = goTermModifier.createNewGOTerm(goTermsFromFile, goToECNumbers, goID, currentCategory.getReactomeName(), GoTermsUpdater.goRefDB);
+					newGOTermsLogger.info("{}\t{}\t{}",dbID,goID,goTermsFromFile.get(goID));
+					if ( ((GONamespace)goTermsFromFile.get(goID).get(GoUpdateConstants.NAMESPACE)).getReactomeName().equals(ReactomeJavaConstants.GO_MolecularFunction) )
+					{
+						newMFLogger.info("{}\t{}\t{}", dbID, goID, goTermsFromFile.get(goID).get(GoUpdateConstants.NAME));
+					}
+					newGoTermCount++;
+				}
+				else
+				{
+					// Try to update each goInstance that has the current GO ID.
+					for (GKInstance goInst : goInstances)
+					{
+						// Compartment is a sub-class of GO_CellularComponent - but the GO namespaces don't seem to account for that,
+						// we we'll account for that here.
+						if (goInst.getSchemClass().getName().equals(currentCategory.getReactomeName()) 
+							|| ( (goInst.getSchemClass().getName().equals(ReactomeJavaConstants.Compartment) || goInst.getSchemClass().getName().equals(ReactomeJavaConstants.EntityCompartment) )
+									&& currentCategory.getReactomeName().equals(ReactomeJavaConstants.GO_CellularComponent) )
+							)
+						{
+							//Now do the update.
+							goTermModifier = new GoTermInstanceModifier(this.adaptor, goInst, this.instanceEdit);
+							goTermModifier.updateGOInstance(goTermsFromFile, goToECNumbers, this.nameOrDefinitionChangeStringBuilder);
+						}
+						else
+						{
+							mismatchCount++;
+							categoryMismatchStringBuilder.append("Category mismatch! GO ID: ").append(goID).append(" Category in DB: ").append(goInst.getSchemClass().getName()).append(" category in GO file: ").append(currentCategory).append("\n");
+							instancesForDeletion.add(goInst);
+						}
+					}
+				}
+			}
+			else if (goTermsFromFile.get(goID).containsKey(GoUpdateConstants.PENDING_OBSOLETION) && goTermsFromFile.get(goID).get(GoUpdateConstants.PENDING_OBSOLETION).equals(true))
+			{
```

Comment:
Regarding the obsoletion of terms, is there anyway to update old ones or is this part of something like a 'total restart' to gene ontology, meaning there isn't a 1:1 way of connecting obsolete terms to new ones?

Reply:
In the old Perl code, obsolete GO terms were logged and deleted. Anything that referred to those terms was modified to refer to the GO terms that were suggested for the obsolete term, in fields such as "alternate", "replaced_by", or "consider". A couple of weeks ago, Peter said that we should not be removing them, but giving the list of obsolete terms to curators so the curators can make a decision on how they are replaced. It sounds like this process of replacing obsolete GO terms is not always simple or straightforward and that doing it automatically is maybe not the best idea.

Reply:
Makes sense, good to know